# COMP0164: Digital Finance
## Group Coursework

If you have set up an **`Anaconda`** environment, don't forget to activate it using:

> conda activate digital_finance_env

Below are the imports that you will need for the coursework:

In [1]:
#Test
import datetime
import numpy as np
import pandas as pd
import scipy as sp

import yfinance as yf
import numpy_financial as npf

import matplotlib.pyplot as plt
%matplotlib inline

## Question 1: Corporate Finance **\[15 marks\]**
In this question, you are asked to evaluate business and investment ideas.

You are tasked with evaluating an investment project proposed by a company called Boogle. During the latest earnings report, Boogle's CFO announced an investment of \$150 million for a new business expansion project. The project is planned to be financed with an \$100 million public offering of a 10-year debt and the remainder with an equity offering. You have collected the information necessary to evaluate this project in Exhibits 1 and 2.

### Exhibit 1: Relevant Information for Analysis
| | |
| --- | --- |
| Equity risk premium | 4.53% |
| Risk-free rate of interest | 3.9% |
| Market value of Boogle's debt | \$1.0 billion |
| Market value of Boogle's equity | \$2.6 billion |
| Boogle's equity beta | 1.4 |
| Boogle's before-tax cost of debt | 9.2% |
| Corporate tax rate | 37.5% |

### Exhibit 2: Estimated Project Financials
|     | Year 1 | Year 2 | Year 3 |
| --- | --- | --- | --- |
| Revenue | 98.7 | 110.3 | 112.6 |
| Operating Costs | 32  | 36  | 38  |
| Depreciation | 16  | 16  | 16  |

a.) Calculate the after-tax weighted average cost of capital of Boogle prior to its new project investment. **\[3 marks\]**

Expected return on equity portion $r_E$ can be calculated by:

$$
r_E = r_f + \beta_E (E[r_m] - r_f)
$$

where 
- $r_f$ is the risk-free rate of interest 
- $\beta_E$ is the equity beta 
- $E[r_m] - r_f$ can be calculated as the equity risk premium

The after-tax weighted average cost of capital (WACC) can be calculated as:

$$
WACC = r_D (1 - T_c) \frac{D}{D + E} + r_E \frac{E}{D + E}
$$

where 
- $r_D$ is the before-tax cost of debt
- $T_c$ is the corporate tax rate
- $D$ is the market value of Boogle's debt
- $E$ is the market value of Boogle's equity
- $r_E$ is the expected return on equity

In [2]:
# Define the variables
equity_risk_premium = 4.53 / 100    # Equity risk premium
risk_free_rate = 3.9 / 100          # Risk-free rate of interest
market_value_debt = 10**9           # Market value of Boogle's debt
market_value_equity = 2.6 * 10**9   # Market value of Boogle's equity
beta_equity = 1.4                   # Boogle's equity beta
expected_return_debt = 9.2 / 100    # Before-tax cost of debt
corporate_tax_rate = 0.375          # Corporate tax rate


In [3]:
def calculate_expected_return_equity(risk_free_rate, equity_beta, equity_risk_premium):
    return risk_free_rate + equity_beta * equity_risk_premium

def after_tax_WACC(r_d, r_e, t_c, d, e):
    return r_d * (1 - t_c) * d / (d + e) + r_e * e / (d + e)


expected_return_equity = calculate_expected_return_equity(risk_free_rate, beta_equity, equity_risk_premium)
wacc = after_tax_WACC(expected_return_debt, expected_return_equity, corporate_tax_rate, market_value_debt, market_value_equity)

print(f"The after-tax weighted average cost of capital of Boogle is: {wacc:.2%}")

The after-tax weighted average cost of capital of Boogle is: 8.99%


In [4]:
# # This is the testing code from previous year
# # Remove it before submitting!!!!!!!!!!!!!!

# def WACC(D, E, r_d, t, r_f, beta, r_p):
#     w_d = D / (D+E)
#     w_e = E / (D+E)
#     r_e = r_f + beta * r_p
#     return w_d * r_d * (1-t) + w_e * r_e

# wacc = WACC(10**9, 2.6*10**9, 0.092, 0.375, 0.039, 1.4, 0.0453)
# print('The weighted average cost of capital is: ', '{0:.4%}'.format(wacc))

b.) Find Boogle's after-tax asset beta prior to the new project. **\[2 marks\]**

In [5]:
# Answer here

beta_e = 1.4
debt = 1.0*10**9
equity = 2.6*10**9
tax_rate = 0.375

def find_beta_a(beta_e, debt, equity, tax_rate):
    return beta_e * (1/(1 + debt*(1-tax_rate)/equity))
# v = debt + equity  # total value of the securities
# beta_a = beta_e * (1 +  debt*(1-tax_rate)/equity)

beta_a = find_beta_a(beta_e, debt, equity, tax_rate)


# def find_beta_e(beta_a, debt, equity, tax_rate):
#     return beta_a * (1 + debt * (1 - tax_rate) / equity)

print(beta_a)
# print(find_beta_e(beta_a, debt, equity, tax_rate))

1.1286821705426355


c.) Assuming the new project has the same asset beta as the Boogle company in b.), find the the project equity beta. **\[2 marks\]**

The project equity beta $ \beta_{pe} $ can be calculated from:

$$
\beta_{pe} = \beta_{asset} \cdot \left[1 + (1 - t) \cdot \frac{D_p}{E_p}\right]
$$

where $ \beta_{asset} $ is the asset beta, $ t $ is the tax rate, $ D_p $ is the project debt, and $ E_p $ is the project equity.

Therefore, we have:

$$
\beta_{pe} = 1.12868 \cdot \left[1 + (1 - 37.5\%) \cdot \frac{100}{50}\right] \\
= 1.12868 \cdot 2.25 \\
= 2.53953
$$

In [6]:
# Answer here
def find_beta_pe(beta_a, t, d_p, e_p):
    return beta_a * (1 + (1 - t) * d_p / e_p)  # Calculate the equity beta (Beta_E)

total = 150 * 10**6  # Total investment
d_p = 100 * 10**6    # Debt portion
e_p = total - d_p    # Equity portion calculated as total investment minus debt
tax_rate = 0.375     # Corporate tax rate (37.5%)

project_beta_pe = find_beta_pe(beta_a, tax_rate, d_p, e_p) # Calculate project equity beta

print(f"Project Equity Beta: {project_beta_pe:.5f}")

Project Equity Beta: 2.53953


In [7]:
# # testing code
# def beta_project(beta_a, t, D_p, E_p):
#     return beta_a * (1+(1-t)*D_p/E_p)

# beta_pe = beta_project(beta_a, 0.375, 100, 50)
# print('The project equity beta is: ', '{0:.4f}'.format(beta_pe))

d.) The formula for project after-tax free cash flow at time t is

$$ FCF = (\text{Revenue}-\text{Operating Costs}-\text{Depreciation})\times(1-\text{Tax Rate})+\text{Depreciation}.$$

Define a Python function to calculate the project FCFs and demonstrate that the after-tax free cash flow generated for the next three years are $47.7 million, $52.4 million, and $52.6 million respectively. **\[2 marks\]**

In [8]:
# Answer here
#Revenues are 98.7, 110.3, 112.6 
#Operating Costs are 32, 36, 38
#Depreciation is 16, 16, 16
#Tax Rate is 0.375
#The initial investment is 150 million
#FCF = (Revenues - Operating Costs - Depreciation) * (1 - Tax Rate) + Depreciation
#The Values from required are stored in the lists below
revenues = [98.7, 110.3, 112.6]
operating_costs = [32, 36, 38]
depreciation = [16, 16, 16]
#The Free Cash Flow for each year is appended to this list so that it can be used later in the next section to find NPV and IRR
FCFs = [-150]
def FCF(revenue, operating_cost, depreciation, tax_rate):
    #The formula mentioned above is used in this function to calculate the Free Cash Flow
    FCF = (revenue - operating_cost - depreciation) * (1 - tax_rate) + depreciation
    return FCF #The function returns the Free Cash Flow
#The for loop is used to calculate the Free Cash Flow for each year
for i in range(len(revenues)):
    #The FCF function is called to calculate the Free Cash Flow for each year
    fcf= FCF(revenues[i], operating_costs[i], depreciation[i], 0.375)
    #The calculated FCF is appended to the FCFs list
    FCFs.append(fcf)
    #The FCF for each year is printed
    print("The FCF for year {i} is ${fcf:.1f} million".format(i=i+1, fcf=fcf))

The FCF for year 1 is $47.7 million
The FCF for year 2 is $52.4 million
The FCF for year 3 is $52.6 million


e.) Find the project NPV and IRR with the next three years after-tax free cash flow given in d.).**\[3 marks\]**

The NPV formula:

$$
NPV = \sum_{t=0}^{n} \frac{FCF_t}{(1 + r)^t} - C_0
$$

Where:
- $C_0$ : the initial investment

The IRR formula:

$$
0 = \sum_{t=0}^{n} \frac{FCF_t}{(1 + IRR)^t}
$$

In [9]:
# Answer here
#The Net Present Value (NPV) is calculated using the numpy_financial library's npf.npv() function
#The IRR is calculated using the numpy_financial library's npf.irr() function

# Ensure that the cell containing the definition of wacc is executed before this cell
npv = npf.npv(wacc, FCFs)

irr = npf.irr(FCFs)

print("The Net Present Value (NPV) of the project is: ${npv:.2f} million".format(npv=npv))
print("The Internal Rate of Return (IRR) of the project is: {irr:.3%}".format(irr=irr))



The Net Present Value (NPV) of the project is: $-21.46 million
The Internal Rate of Return (IRR) of the project is: 0.899%


Assume the following is a dividend-paying stock.

### Stock:
Atat Steel is in the steel manufacturing sector with a required rate of return of 7.35%. You estimate that if the economy is booming, the company’s current annual dividend of $0.8 per share will grow 10.5% a year for the next four years and then stabilize at a 3.5% growth rate a year indefinitely. However, if the economy falls into a recession, then Atat Steel will not likely experience the elevated 10.5% short-run growth and instead will grow by 3.5% indefinitely.

f.) Use the discount dividend method and find the current value of Atat Steel stock under both economic conditions. **\[3 marks\]**

The **booming economy scenario**: in order to find the dividened for the first 4 years we use Multi - Staged Dividened Discount Model with the formula

$$
P_0 = \sum_{t=1}^H \frac{D_t}{(1 + r)^t} + \frac{D_H (1 + g_2)}{(r - g_2)(1 + r)^H}
$$

$$ 
Dt = D(t-1) * (1 + g_1)^t 
$$


Where:
- $P_0$ : Current price of the stock.
- $D_0$ : Current dividend.
- $g_1$ : Short-term growth rate.
- $g_2$: Long-term growth rate.
- $r$: Required rate of return.
- $H$ : Number of boom years

The the **recession** scenario: in order to find the dividened the Constant Growth Dividened Method is used as demonstrated below.
$$
P_0 = \frac{D_0 \times  (1 + g_2)}{(r - g_2)}
$$

Where:
- $P_0$: Current price of the stock.
- $D_0$: Current dividend.
- $g_2$: Constant growth rate.
- $r$: Required rate of return.




In [10]:
# Answer here 

#Initialise the parameters for both scenarios
first_dividend = 0.8
required_rate_of_return = 0.0735
growth_rate = 0.035

#Function to find the dividends for the next year when there is growth
def Next_year_dividend(dividend, growth_rate):

    return dividend * (1 + growth_rate)

#Function to find the perpetual dividend
def perpetual_dividend(dividend, growth_rate, required_rate_of_return):
    return dividend * (1 + growth_rate) / (required_rate_of_return - growth_rate)

#Function to find the dividends for each year during the growth period
def Dividends_growth_period(first_dividend, boom_years):
    boom_dividends = []
    for i in range(boom_years):
        #The first year dividend is calculated using the formula: Dividend * (1 + g)
        if i == 0:
            next_year_dividend = Next_year_dividend(first_dividend, growth_rate_boom)
        #The subsequent year dividends are calculated using the formula: Previous year Dividend * (1 + g)^the current year
        else:
            next_year_dividend = Next_year_dividend(boom_dividends[-1], growth_rate_boom)
        boom_dividends.append(next_year_dividend)
    return boom_dividends

#Function to find the NPV of the dividends during the growth period
def NPV_dividends_growth(boom_dividends, required_rate_of_return):
    boom_dividends_npv = 0
    for i in range(len(boom_dividends)):
        #The NPV of the dividends is calculated using the formula: Dividend of year t / (1 + r) ^ t
        boom_dividends_npv += boom_dividends[i] / ((1 + required_rate_of_return) ** (i + 1))
    return boom_dividends_npv

#This section is for the Booming economy:

#Initialise the parameters for the booming economy
years_boom = 4
growth_rate_boom = 0.105

#Calculate the dividends for each year during the booming economy
boom_dividends = Dividends_growth_period(first_dividend, years_boom)

#Find the NPV for the first 4 years of the booming economy:
boom_dividends_npv = NPV_dividends_growth(boom_dividends, required_rate_of_return)

#Find the perpetual growth rate for the dividends during the booming economy
perpetual_boom = perpetual_dividend(boom_dividends[-1], growth_rate, required_rate_of_return)

#Find the NPV of the perpetual growth rate for the dividends during the booming economy
perpetual_boom_npv = perpetual_boom/((1+required_rate_of_return)**years_boom)

#Calculate the terminal value of the company during the booming economy

P0_boom = boom_dividends_npv + perpetual_boom_npv

print("The price of the stock in the case of booming economy is: ${P0_boom:.2f}".format(P0_boom=P0_boom))

#This section is for the Recession economy:

P0_recession = perpetual_dividend(first_dividend, growth_rate, required_rate_of_return)

print("The price of the stock in the case of recession economy is: ${P0_recession:.2f}".format(P0_recession=P0_recession))



The price of the stock in the case of booming economy is: $27.59
The price of the stock in the case of recession economy is: $21.51


In [ ]:
#Delete this cell before submitting!!!!!!!!!
#Test case for our code and we get the same result using the take home exam code
"""def compute_npv(CF, r):
    npv = CF[0]
    for i in range(1,len(CF)):
        npv += CF[i] / (1+r)**i
    return npv

D0 = 0.8
g1 = 0.105
g2 = 0.035
r = 0.0735

D1 = D0 * (1+g1)
D2 = D0 * (1+g1)**2
D3 = D0 * (1+g1)**3
D4 = D0 * (1+g1)**4
D5 = D0 * (1+g1)**4 * (1+g2)

P4 = D5 / (r-g2)

P0_booming = compute_npv([0, D1, D2, D3, D4+P4], r)
print('The current value when the economy is booming: ', '{0:0.2f}'.format(P0_booming))

P0_recession = D0 * (1+g2) / (r-g2)
print('The current value when the economy falls into a recession: ', '{0:0.2f}'.format(P0_recession))"""

The current value when the economy is booming:  27.59
The current value when the economy falls into a recession:  21.51


## Question 2: Bonds and Fixed Income **\[16 marks\]**
In this question, you are given background information on some bonds and you tasked with calculating crucial data about them.

### Exhibit 3: Current Par Rates and Spot Rates
| Maturity | Par Rate (Annual) | Spot Rate (Annual) |
| --- | --- | --- |
| 1 year | 2.50% | 2.50% |
| 2 years | 3.02% | 3.00% |
| 3 years | 3.53% | 3.50% |
| 4 years | 3.98% | 4.00% |
| 5 years | 4.42% | ? |

### Exhibit 4: Information for Selected Bonds
| Bond Name | Maturity | Coupon (Annual) | Type of Bond |
| --- | --- | --- | --- |
| Bond A (Face value $1,000) | 3 years | 6.78% | Option-free |
| Bond B | 3 years | 4.30% | Callable at par on start of year 1 and tear 2 |
| Bond C | 3 years | 4.30% | Putable at par on start of year 1 and year 2 |

### Exhibit 5: Binomial Interest Rate Tree, based on an estimated interest rate volatility of 10%, where ‘u’ represents an up move and ‘d’ represents a down move.
| Year 0 | Year 1 | Year 2 |
| --- | --- | --- |
| 2.3400% (r) | 3.5930% (ru node) | 4.6540% (ruu node) |
|  | 2.9417% (rd node) | 3.8043% (rud node) |
|  |  | 3.1254% (rdd node) |

a.) Based on Exhibit 3, find the five-year spot rate. **\[2 marks\]**

To calculate the five-year spot rate, we use the following formula

$$
P = \sum_{t=1}^{5} \frac{CF_t}{(1 + S_t)^t} + \frac{P}{(1 + S_5)^5}
$$

Where:
- $P$: Price of the bond (assume $P = 100 $ for par value).
- $CF_t$: Annual coupon payment ($(C = 100 \times \text{Par Rate}$)).
- $S_t$: Known spot rates for $(t = 1, 2, 3, 4)$.
- $S_5$: The unknown spot rate for 5 years.

Steps:
$$
P = \frac{CF_1}{(1+S_1)}+ \frac{CF_2}{(1+S_2)^2}+ \frac{CF_3}{(1+S_3)^3}+ \frac{CF_4}{(1+S_4)^4}+ \frac{CF_5 + P}{(1+S_5)^5}
$$

$$
100 = \frac{CF_1}{(1+S_1)}+ \frac{CF_2}{(1+S_2)^2}+ \frac{CF_3}{(1+S_3)^3}+ \frac{CF_4}{(1+S_4)^4}+ \frac{CF_5 + 100}{(1+S_5)^5}
$$

Where:

- $CF$: $0.0442 \times\ 100$
$$
100 = \frac{(4.42)}{(1+0.025)}+ \frac{(4.42)}{(1+0.03)^2}+ \frac{(4.42)}{(1+0.035)^3}+ \frac{(4.42)}{(1+0.04)^4}+ \frac{(4.42 + 100)}{(1+S_5)^5}
$$

$$
 (1+S_5)^5 = 1.247



In [12]:
# Answer here
#The parameters for the calculation are defined
spot_rate_list = [0.025, 0.03, 0.035, 0.04]

par_rate = 0.0442
#Assuming a face value of $100 to calculate the five-year spot rate
p = 100

def find_spot_rate(p, par_rate, spot_rate_list):
    #The spot rate for each year is calculated usiing the formula: (p*par) / (1 + spot_rate) ^ t
    total_spot_rate_vals = 0
    for i in range(len(spot_rate_list)):
        total_spot_rate_vals += (p * par_rate)/((1 + spot_rate_list[i])**(i+1))
    print(total_spot_rate_vals)
    #The formula (1+S5)^5 = (100+4.42)/(100 - 16.24329032372646)
    spot_rate_val_power = (p + (par_rate * p))/(p - total_spot_rate_vals)
    print(spot_rate_val_power)
    #The spot rate for the 5th year is calculated using S_5 = ((100 + 4.42)/(100 - total_spot_rate_vals)) ^ (1/5) - 1
    spot_rate = spot_rate_val_power**(1/(len(spot_rate_list)+1)) - 1

    return spot_rate

five_year_spot_rate = find_spot_rate(p, par_rate, spot_rate_list)

print("The five-year spot rate is: {five_year_spot_rate:.3%}".format(five_year_spot_rate=five_year_spot_rate))





16.24329032372646
1.2467060896206614
The five-year spot rate is: 4.509%


In [13]:
#Delete this cell before submitting!!!!!!!!!
#Test case for our code and we get the same result using the take home exam guy code
"""def compute_spot_rate(spot_rate_list, par_rate):
    f = 1
    cf = 1 * par_rate
    for i in range(len(spot_rate_list)):
        f -= cf / ((1+spot_rate_list[i])**(i+1))
    print(f)
    f = (1+cf) / f
    print(f)
    s = f**(1/(len(spot_rate_list)+1)) - 1
    return s
spot_rate_list = [0.025, 0.03, 0.035, 0.04]

par_rate = 0.0442
s_5 = compute_spot_rate(spot_rate_list, par_rate)
print('The five-year spot rate is: ', '{0:.4%}'.format(s_5))"""


"def compute_spot_rate(spot_rate_list, par_rate):\n    f = 1\n    cf = 1 * par_rate\n    for i in range(len(spot_rate_list)):\n        f -= cf / ((1+spot_rate_list[i])**(i+1))\n    print(f)\n    f = (1+cf) / f\n    print(f)\n    s = f**(1/(len(spot_rate_list)+1)) - 1\n    return s\nspot_rate_list = [0.025, 0.03, 0.035, 0.04]\n\npar_rate = 0.0442\ns_5 = compute_spot_rate(spot_rate_list, par_rate)\nprint('The five-year spot rate is: ', '{0:.4%}'.format(s_5))"

b.) Assuming the law of one price, use Exhibit 3 to calculate the compounded forward rate (annualy compounded) of a one-year loan starting in three years. **\[1 mark\]**

$$
f_(34) = \frac {(1 + S_4)^4}{(1 + S_3)^3} -1
$$
Where
- $S_4$: 4.00%
- $S_3$: 3.50%
$$
f_(34) = \frac {(1 + 0.04)^4}{(1 + 0.035)^3} -1
$$


In [36]:
# Answer here

#The parameters for the calculation are defined
spot_rate_year_3 = 0.035
spot_rate_year_4 = 0.04

#The spot rate for the 3rd year is 3.5% and the spot rate for the 4th year is 4% based on Exhibit 3

def find_forward_rate(spot_rate_year_3, spot_rate_year_4):
    #The forward rate for the 4th year is calculated using the formula: (1 + spot_rate_year_4) ^ 4 / (1 + spot_rate_year_3) ^ 3 - 1
    forward_rate = ((1 + spot_rate_year_4) ** 4 / (1 + spot_rate_year_3) ** 3) - 1

    return forward_rate

forward_rate = find_forward_rate(spot_rate_year_3, spot_rate_year_4)

print("The forward rate for a one-year loan made in three years is: {spot_rate_one_year_loan:.4%}".format(spot_rate_one_year_loan=forward_rate))

The forward rate for a one-year loan made in three years is: 5.5145%


c.) Given spot rates for one-, two-, and three-year zero bonds, how many forward rates can be calculated? Please list the compounded forward rates (annualy compounded) that can be calculated and briefly explain your answer. **\[3 marks\]**

Forward rate formula is: 

$$
RF = \frac {R2 \times T2 - R1 \times T1}{T2 - T1}
$$

In [ ]:
# Answer here
"""
The first forward rate can be calcualted for a one year loan starting in year 0 (f0,1 = S1)
The second forward rate can be calcualted for a one year loan starting year 1 (f1,2)
The third forward rate can be calcualted for a one year loan starting year 2 (f2,3)
A fourth forward rate can be calcualted for a two year loan spanning for 2 years starting in year 1 until year 3 (f1,3)

f(0,1) = S1 = 0.025
f(1,2) = (1+S2)^2 / (1+S1)^1 - 1 
f(2,3) = (1+S3)^3 / (1+S2)^2 - 1
f(1,3) = (1+S3)^3 / (1+S1)^1 - 1

"""

d.) Find the yield to maturity for Bond A. **\[4 marks\]**

In order to calculate yield to maturity of for Bond A, the following formula is used: 

$$
PV = \sum_{t = 1}^n \frac{C}{(1 + y)^t} + \frac{P}{(1+y)^n}
$$

Where
- $PV$ : Present value or price of the bond 
- $P$ : Face value of the bond (In the case of Bond A = $1000)
- $C$ : Annual coupon payments (in the case of Bond A: $1000 \times 6.78% = $67.8)
- $y$ : Yield to maturity of the bond
- $n$ : Number of periods to maturity (In case of Bond A this is 3 years)

$$
1093.14 = \frac{67.8}{(1 + 0.025)} + \frac{67.8}{(1 + 0.03)^2} + \frac{67.8}{(1 + 0.035)^3} + \frac{1000}{(1 + 0.035)^3}
$$
$$
1093.14 = \frac{67.8}{(1 + y)} + \frac{67.8}{(1 + y)^2} + \frac{67.8}{(1 + y)^3} + \frac{1000}{(1 + y)^3}
$$

In [ ]:
# Answer here
#Defining the spot rates for Bond A
face_val = 1000
coupan_rate = 0.0678
maturity = 3
spot_rate_lst = [0.025, 0.03, 0.035]
#Defining the function to find the present value of the bond
def find_PV(face_val, coupan, maturity, spot_rates):
    #The present value of the bond is calculated using the formula: (C / (1 + r) ^ t) + (F / (1 + r) ^ t)
    coupans = coupan * face_val

    PV = 0
    for i in range(len(spot_rates)+1):
        #The present value of the coupan is calculated for each year and also the present value of the face value for the last year is calculated then all are added to get the present value of the bond
        if i == len(spot_rates)-1:
            PV += (face_val + coupan) / ((1 + spot_rates[i-1]) ** (i))
            break
        else:
            PV += coupans / ((1 + spot_rates[i]) ** (i+1))
    return PV
PV = find_PV(face_val, coupan_rate, maturity, spot_rate_list)

#Define function to find the yield to maturity

def Find_ytm(PV, face_val, coupan, maturity):
    #Find the values for the bond to calculate the yield to maturity
    vals = [-PV]
    #Calculate the coupan for each year 
    coupans = coupan * face_val
    #Append the coupan for each year to the list of values and the face value for the last year
    for i in range(maturity):
        if i == maturity-1:
            vals.append(face_val + coupans)
        else:
            vals.append(coupans)
    #The yield to maturity is calculated using the numpy_financial library's npf.irr() function
    ytm = npf.irr(vals)

    return ytm

ytm = Find_ytm(PV, face_val, coupan_rate, maturity)
print("The yield to maturity of Bond A is: {ytm:.3%}".format(ytm=ytm))

The yield to maturity of Bond A is: 3.456%


e.) Based on Exhibit 5, assume an equal probability of interest rate going up and down at each node. Calculate the value of Bond B and Bond C with the binomial tree model. **\[3 marks\]**

In [17]:
# Answer here

f.) All else being equal, explain the effect of a fall in interest rates on Bond B and Bond C. **\[2 marks\]**

In [18]:
# Answer here

g.) All else being equal, which bond is most likely to increase in value if interest rate volatility is 15% rather than 10%? Briefly explain your answer. (Hint: consider the value of options) **\[1 mark\]**

In [19]:
# Answer here

## Question 3: Forward and Futures Contracts **\[12 marks\]**
In this question you will be working with futures and forward contracts

Consider a stock that is trading at $100 today. The stock does not generate income/pay dividends. The stock is traded in a well-functioning market with no transaction costs and no restrictions on short sales. Both borrowing and lending can be done in unlimited amounts at the 2% risk-free rate.

a.) What is the difference between forward contracts and futures contracts (Answers should be no longer than 200 words) **\[4 marks\]**

**Answer**:

Forward contracts are agreements between two parties to buy or sell an asset at a future date for a price agreed upon today. This is an over-the-counter agreement so they are privately negotiated between two parties and the contract terms are customised to fit the needs. Since each contract is tailor-made, it's not standardised, giving flexibility but less liquidity. This comes with higher credit risk because there is no intermediary to guarantee the transaction. The forward contracts are mainly used for hedging.

Future contracts are standardised agreements traded on an exchange. It is marked-to-market daily, meaning gains and losses are settled daily so the credit risk is reduced. The terms of this contract are set by an exchange and the contract is fully standardised, ensuring high liquidity and ease of trading. The credit risk is reduced as mentioned previously, but also because there is a margin system in place to manage risk. Future contracts can be used for both hedging and speculative purposes.


b.) Consider a futures contract on the stock with a maturity of one year. Suppose that the futures price is currently at $97. Are the futures fairly priced? Describe an arbitrage strategy that would allow you to make a riskless profit. **\[2 marks\]**

**Answer here:**

No this is not fairly priced, as the future price can be expected as approximately $102. Here is the calculation.

To calculate the future value, the following formula can be used:

$$
F = S \cdot e^{rT}
$$

where \( S \) is the current price of the asset, \( r \) is the risk-free interest rate and \( T \) is the time to maturity in years.

However, I have simplified this into the following equation by using Taylor series, as in this question it is assumed that the maturity of the stock is only one year and the formula above might compute a large error in a short time to maturity. The following formula can be used instead:

$$
F = S \cdot (1 + r)^T
$$

Applying this question into this formula,

$$
F = 100 \cdot (1 + 0.02)^1 = 102
$$

The expected future price is $102.


**To make a riskless profit:**

1. Buy the futures contract at $97, meaning we agree to buy the stock after a year for $97. 

2. Short-sell the stock today at the spot price, which in this case $100.

3. Invest the proceeds ($100) from the short sale at the risk-free rate. 

This will gain the total riskless profit of approximately $5 per share. This is possible by using the $102 which gained from investing $100 at the 2% of risk-free rate and spending that money to buy the stock at $97. Therefore:
$$
102 - 97 = 5
$$

The riskless profit will be $5 per share. 

c.) Same as question b) but suppose that the futures price is currently at $103. Describe your arbitrage strategy. **\[2 marks\]**

In [22]:
# Answer here

d.) Consider a stock with the market value of $76 that pays 5% dividend yield over the next year. Given the risk-free rate of return is 3%, and the initial forward price is $73, what is the present value of the forward contract with one year maturity? **\[2 marks\]**

In [23]:
# Answer here

e.) Consider the same conditions as in part d.), but with the dividend of $3, what is the present value of the forward contract? **\[2 marks\]**

In [24]:
# Answer here

# Question 4: Options **\[16 Marks\]**
In this question, you will be working with options.

Suppose that you hold a long position on a European call option that has an underlying asset price of $45.34, strike price of $44, risk-free rate of 1.5%, 20% annualized volatility, and a half-a-year left to maturity. The underlying asset does not have any investment yield.

a.) Value this call option. **\[1 mark\]**

In [25]:
# Answer here

b.) Based on the Black-Scholes-Merton model, describe a portfolio that replicates the call option’s payoff. **\[1 mark\]**

In [26]:
# Answer here

c.) Define a function to price the option with the binomial tree method. The function should take the number of steps (n) as one of the inputs. You should NOT use list comprehension in the function. **\[3 marks\]**

In [27]:
# Answer here

d.) By setting n = 10, 50 and 100, compare and comment on the results under the two methods. **\[2 marks\]**

In [28]:
# Answer here

When buying two calls with the exercises price of $x_{1}$ and $x_{3}$ and selling two calls with the exercise price of $x_{2}$, where
$$x_{2} = \frac{x_{1}+x_{3}}{2},$$
with the same maturity for the same stock, we call it a butterfly. Consider the following call options for the stock that trades at $57.03:
| Option Name | Strike Price | Call Premium (Price) |
| --- | --- | --- |
| Call Option 1 | 50 | 10 |
| Call Option 2 | 55 | 7 |
| Call Option 3 | 60 | 5 |

e.) Create a graphical representation of the butterfly strategy’s payoff. **\[2 marks\]**

In [29]:
# Answer here

f.) Why might an investor enter into such a strategy? **\[2 marks\]**

In [30]:
# Answer here

g.) Use binomial model (set some arbitrary high number of iterations) to price the european butterfly option from part e.), assuming it has 1 year to maturity, the risk-free rate is 1.5% and the annualized volatility of the underlying stock is 15%. **\[5 marks\]**

In [31]:
# Answer here

# Question 5: Case Study **\[14 marks\]**
Using the data sources and tools introduced in Workshop 3, along with any relevant material from the lectures, analyze the current state and future outlook of a selected country/economy from a macroeconomic and financial perspective. As a primary suggestion, groups may consider analyzing the United States economy, as most available free data sources focus on this country. However, groups are welcome to choose any country/economy they wish to analyze. Groups MUST include figures and/or tables from various data sources, as well as research reports or studies that justify and support the analysis conducted. The report should be approximately 600/700 words, excluding the bibliography, captions for tables and figures, and any numbers included in tables.

# Question 6: Case Study **\[17 marks\]**
Using the data sources and tools introduced in Workshop 3, along with any relevant material from the lectures, analyze the current state and future outlook of a selected publicly traded company (stock) included in a major stock market index, such as the S&P 500 or Russell 2000, among others. The company may be located in a different country than the one chosen in Question 5. As a primary suggestion, groups may consider analyzing a company from the “Magnificent 7”: Apple, Microsoft, Alphabet (Google), Amazon, Nvidia, Tesla, or Meta (Facebook). You should include, at least, a brief introduction to the company (i.e., a description of its main business) and a fundamental analysis.

Groups MUST include figures and/or tables from a variety of data sources, as well as research reports or studies that justify and support the analysis conducted. The report should be approximately 600/700 words, excluding the bibliography, captions for tables and figures, and any numbers included in tables.

You are not limited to analyzing an “excellent” publicly company/stock; you may also consider a poorly performing company whose stock price may decline, making a short position potentially profitable.

Each group, working as financial analysts, must provide a final conclusion based on their analysis by assigning a rating to the company. The rating scale is as follows:
- (5) Strong Sell
- (4) Sell
- (3) Neutral
- (2) Buy
- (1) Strong Buy

Note that calculating a (target) price for the company is not required; simply provide a recommendation based on thorough research and analysis.